src/allmydata/test/cli/wormholetesting.py

```diff
@@ -184,7 +183,7 @@ def allocate_code(self, wormhole: IWormhole, code: Optional[WormholeCode]) -> Wo
 
         return code
 
-    def wait_for_wormhole(self, code: WormholeCode) -> Awaitable[_MemoryWormhole]:
+    def wait_for_wormhole(self, code: WormholeCode) -> Deferred[_MemoryWormhole]:
```

Comment:
My Python isn't strong enough to know - are these equivalent?

Reply:
I think it is equivalent, but if we ever end up using ```--strict``` or I think ```--disallow-untyped-defs``` flag in our mypy config mypy does not like this and will complain that return type is not the same.

Reply:
`Awaitable` is a protocol - kind of like an interface or an abstract base class.  You cannot have a value the type of which is `Awaitable` but you can have a value that is an "instance" of it  Protocols are "structural" so a value is considered an "instance" of `Awaitable` if it has attributes and methods that "agree" with those defined by `Awaitable`.  In this case, `Awaitable` defines one method, `__await__`, that takes only a `self` argument.  Any value with such a method will be considered an instance of `Awaitable`.  `Deferred` is one type the instances of which qualify.

In general, it is fine for a method that returns a `Deferred` to be annotated as returning an `Awaitable` because all values of type `Deferred` are instances of `Awaitable`.  However, in this specific case `_WormholeApp.wait_for_wormhole` is used by `_WormholeServerView.wormhole_by_code`.  The latter is annotated as returning `Deferred[IWormhole]` and returns whatever the former returns.

It is not permissible to return an `Awaitable` where the return type has been annotated as `Deferred` because not all values of `Awaitable` are values of `Deferred`.  Switching `wait_for_wormhole`'s return annotation makes it consistent with `wormhole_by_code` and fixes a type error where an `Awaitable` was being used as a `Deferred`.

The fix could just as easily gone the other way - switching `wormhole_by_code` from `Deferred` to `Awaitable`.  Doing it that way allows the more general `Awaitable` type to propagate outwards to users of `wormhole_by_code`.  And, in fact, there is at least one user that acts as if the return value really is a `Deferred` and not an `Awaitable` and so it would also need to be fixed (somehow).  Fixing this by propagating the type change "down" the dependency tree ends up being easier in this case.

There isn't any real sense in which one of these changes is more correct than the other.  They both produce something consistent and decideable.  But the way it is fixed in the branch right now looks like the smaller change to me, which is as reasonable a tie-breaker as anything, lacking other more substantial considerations.

Reply:
I think one way using `Awaitable` is preferable is if we ever want to upgrade to `async def` ... right? (Then, one can "slowly" upgrade things towards `async def` while complying with the type-signature -- i.e. not having to use `ensureDeferred()` all over the place).

Reply:
Right, this is probably true.  Having `Deferred` in the annotation here doesn't directly impede that (right now, the caller could already `await ...` this and everything would be fine).  The advantage `Awaitable` has over `Deferred` as the return annotation is that the _only_ thing you can do with an `Awaitable` is await it (or pass it to something like `ensureDeferred` - which just awaits for you and translates the interface) whereas while it's a `Deferred` you can still use `addCallback` instead.

I fully expect that eventually this API (and probably all similar) will become `async def ... -> X:` (where `Awaitable` is implied by `async def`) but even having `Deferred` annotations on non-async def functions is a good step towards that since it will let mypy help us make that transition with fewer bugs.